vscode/src/circuit.ts

```diff
@@ -0,0 +1,134 @@
+
+export async function showCircuitCommand(
+  extensionUri: Uri,
+  operation: IOperationInfo | undefined,
+) {
+  const associationId = getRandomGuid();
+  sendTelemetryEvent(EventType.TriggerCircuit, { associationId }, {});
+
+  const compilerWorkerScriptPath = Uri.joinPath(
+    extensionUri,
+    "./out/compilerWorker.js",
+  ).toString();
+
+  const editor = window.activeTextEditor;
+  if (!editor || !isQsharpDocument(editor.document)) {
+    throw new Error("The currently active window is not a Q# file");
+  }
+
+  sendMessageToPanel("circuit", true, undefined);
+
+  let timeout = false;
+
+  // Start the worker, run the code, and send the results to the webview
+  const worker = getCompilerWorker(compilerWorkerScriptPath);
+  const compilerTimeout = setTimeout(() => {
+    timeout = true;
+    sendTelemetryEvent(EventType.CircuitEnd, {
+      associationId,
+      reason: "timeout",
+      flowStatus: UserFlowStatus.Aborted,
+    });
+    log.info("terminating circuit worker due to timeout");
+    worker.terminate();
+  }, compilerRunTimeoutMs);
+  let title;
+  let subtitle;
+  const targetProfile = getTarget();
+  const sources = await loadProject(editor.document.uri);
+  if (operation) {
+    title = `${operation.operation} with ${operation.totalNumQubits} input qubits`;
+    subtitle = `${getTargetFriendlyName(targetProfile)} `;
+  } else {
+    title = basename(editor.document.uri.path) || "Circuit";
+    subtitle = `${getTargetFriendlyName(targetProfile)}`;
+  }
+
+  try {
+    sendTelemetryEvent(EventType.CircuitStart, { associationId }, {});
+    const circuit = await worker.getCircuit(sources, targetProfile, operation);
+    clearTimeout(compilerTimeout);
+
+    const message = {
+      command: "circuit",
+      circuit,
+      title,
+      subtitle,
+    };
+    sendMessageToPanel("circuit", false, message);
+
+    sendTelemetryEvent(EventType.CircuitEnd, {
+      associationId,
+      flowStatus: UserFlowStatus.Succeeded,
+    });
+  } catch (e: any) {
+    log.error("Circuit error. ", e.toString());
+    clearTimeout(compilerTimeout);
+
+    const errors: [string, VSDiagnostic][] =
+      typeof e === "string" ? JSON.parse(e) : undefined;
+    let errorHtml = "There was an error generating the circuit.";
+    if (errors) {
+      errorHtml = errorsToHtml(errors);
+    }
+
+    if (!timeout) {
+      sendTelemetryEvent(EventType.CircuitEnd, {
+        associationId,
+        reason: errors && errors[0] ? errors[0][1].code : undefined,
+        flowStatus: UserFlowStatus.Failed,
+      });
+    }
+
+    const message = {
+      command: "circuit",
+      title,
+      subtitle,
+      errorHtml,
+    };
+    sendMessageToPanel("circuit", false, message);
+  } finally {
+    log.info("terminating circuit worker");
+    worker.terminate();
+  }
+}
+
+function errorsToHtml(errors: [string, VSDiagnostic][]): string {
+  let errorHtml = "";
+  for (const error of errors) {
+    const uri = Uri.parse(error[0]);
```

Comment:
What is at element 0? Can you add some comments as to what these commands are you're embedding into the html.

Reply:
Added comments, and in the process discovered a bug handling project-level errors. Thanks!